From 8116db8cfa8edd1c186a57bfd447da35a013f243 Mon Sep 17 00:00:00 2001
From: Julien Grall <julien.grall@citrix.com>
Date: Wed, 2 Sep 2015 18:20:53 +0100
Subject: [PATCH] hw/block: xen-disk: WORKAROUND: disable batch map when using
 64KB page

The gntdev driver is not able to handle correctly the mapping of more
than 1 grant when the Linux page size is 64KB.

This is because the grant code is relying on having one Linux page per
Grant (which is 4KB). So the grant won't be contiguous in the final
mapping.

For now, always map the grant one by one when the Xen page size is
different than the Linux page size.

Signed-off-by: Julien Grall <julien.grall@citrix.com>
---
 hw/block/xen_disk.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/block/xen_disk.c b/hw/block/xen_disk.c
index d90f2cd..9f89875 100644
--- a/hw/block/xen_disk.c
+++ b/hw/block/xen_disk.c
@@ -755,7 +755,7 @@ static void blk_alloc(struct XenDevice *xendev)
     QLIST_INIT(&blkdev->finished);
     QLIST_INIT(&blkdev->freelist);
     blkdev->bh = qemu_bh_new(blk_bh, blkdev);
-    if (xen_mode != XEN_EMULATE) {
+    if (xen_mode != XEN_EMULATE && XC_PAGE_SIZE == getpagesize() ) {
         batch_maps = 1;
     }
     if (xc_gnttab_set_max_grants(xendev->gnttabdev,
-- 
2.1.4

